### PR TITLE
Unify lengthof/string_lengthof -like macros

### PR DESCRIPTION
**Pull request by am11**
* Move `src/native/common` to `src/native/minipal`
* Define `_countof` and `_strcountof`.
* Replace all `_countof`, `lengthof`, `NumItems`, `ARRAY{_}SIZE` like macros with `_countof`.
* Replace all `lengthof` and `StrLen` like macros with `_strcountof`.
    * Also replaced `_countof() - 1` patterns to `_strcountof()` in few places.

In second commit, ran editor macro to cleanup trailing whitespaces in files touched by the first commit.

### CODE REVIEW

**Comment by AraHaan** [comment]
What is ``MENIPAL_*``?

**Comment by am11** [comment]
> What are the `MENIPAL_*` macros used here defined as?

There were about 32 definitions of those macros repo-wide, this patch moves the definitions in a single place:
https://github.com/dotnet/runtime/blob/ad9e7265bf8dbc16d0bf14d55122aba887a28297/src/native/minipal/utils.h#L7-L10

**Comment by am11** [comment]
cc @jkotas

**Comment by jkotas** [comment]
What is the most common name used for this macro in the universe? Would it look better to standardize on that name without any prefix?

**Comment by vargaz** [comment]
There is no common name i think, mono uses the name from glib.

**Comment by SingleAccretion** [comment]
Note that the Jit also uses an [`ArrLen`](https://github.com/dotnet/runtime/blob/0818df12f8c6ffd19f38b10ee5ab43ccd486e537/src/coreclr/jit/utils.h#L34-L38) function for this. I will say that if it were for me to decide, I'd replace all the uses of `_countof` in the Jit with it, but, heh, it is not :).

**Comment by am11** [comment]
I don't have strong opinions about the name (naming stuff is hard), so whatever is decided I'm fine with it; now that find & replace after this patch is much easier. ;)

Prefix convention (used heavily in mono) helps identifying where the macro of function definitions are coming from. Recently, we prefixed functions with `minipal_`. I followed the same here for macros to stay consistent that way.

**Comment by am11** [comment]
@SingleAccretion, had to add `constexpr` specifier to it, but I have now used `ArrLen` in `jit/`. The exception is `jitstd/` subdir; which is also used outside the JIT and cannot resolve all dependencies of utils.h (without compiler.h present).

@jkotas, the most common name is `_countof`, do you think we should standarize on that or keep it prefixed?

**Comment by jkotas** [comment]
> @jkotas, the most common name is _countof, do you think we should standarize on that or keep it prefixed?

Yes, it sounds good to me. I think the prefix is overkill for this one. `_countof` is de-facto standard on Windows already since it is defined in Windows SDK (https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/countof-macro).

**Comment by am11** [comment]
Renamed `MINIPAL_{LENGTHOF, STRLEN}` to `_countof` and `_strcountof` respectively.
CI test failures are unrelated to changes.

**Comment by vargaz** [comment]
That doesn't match the mono coding conventions and looks very much like some kind of compiler intrinsic.

**Comment by jkotas** [comment]
@vargaz What would be your preferred name?

**Comment by vargaz** [comment]
Not sure this particular macro needs to be unified between the runtimes, both of them already has their own version.

**Comment by jkotas** [comment]
I hope that we will share more code between the runtimes over time. I think it is important that we get the details like these unified to something reasonable. Do you have an opinion what should be the name to standardize on in the code shared between runtimes for this one?

**Comment by vargaz** [comment]
Imho the problem is that the runtimes have different coding conventions, so not sure a macro name can fit both equally well. I'd prefer the traditional uppercase macro naming style.

**Comment by am11** [comment]
In src/native/eventpipe is another example of code shared between runtimes, where we have function, as well as the macro names prefixed and macros are upper cased. coreclr uses many different conventions (prefix, no prefix, upper, lower, mixed casing) for macros so there is no problem in changing it to whichever name is suitable.

**Comment by jkotas** [comment]
> I'd prefer the traditional uppercase macro naming style.

Would `ARRAY_SIZE` work?

File: src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h [hunk 1]
```diff
@@ -15,7 +15,6 @@
 #include "clrversion.h"
 
 #undef EP_ARRAY_SIZE
```
Comment (jkotas) [inline src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:17]: Can this undef be deleted?

File: src/libraries/Native/AnyOS/System.IO.Compression.Native/entrypoints.c [hunk 2]
```diff
@@ -1,7 +1,8 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#include <common/entrypoints.h>
+#include <minipal/entrypoints.h>
```
Comment (jkotas) [inline src/libraries/Native/AnyOS/System.IO.Compression.Native/entrypoints.c:4]: Can `entrypoints.h` include `minipal/utils.h`? It looks like one always needs `minimap/utils.h` when you use `entrypoints.h`.

**Comment by jkotas** [comment]
Based on the feedback, here is my proposal to land this PR:
- Rename `_countof`/`_strcountof` to `ARRAY_SIZE` and `STRING_LENGTH`. I think it is reasonable convention to have macros in all caps and I agree that `_countof` looks more like a compiler intrinsic.
- Keep using `G_N_ELEMENTS` in mono since it is used fairly consistently there and follows the Mono coding style.

@am11 If there is no push back in next two days, could you please implement these change? I am sorry about the back and forth, and thank you for your work on various cleanups!

File: src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h (hunk 1 above)
Comment (am11) [reply src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:17]: Yup, it was a leftover. Next batch of changes are currently in [`feature/minipal2` branch](https://github.com/am11/runtime/tree/feature/minipal2). I'll wait for more feedback and push them to this branch later.

**Comment by vargaz** [comment]
Sounds good

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**